lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts

```diff
@@ -55,28 +55,30 @@ export abstract class PnpmLockfileParser {
     return depsRoot;
   }
 
-  public extractPackages(): NormalisedPnpmPkgs {
+  public extractPackages() {
     const packages: NormalisedPnpmPkgs = {};
     Object.entries(this.packages).forEach(
       ([depPath, versionData]: [string, any]) => {
         // name and version are optional in version data - if they don't show up in version data, they can be deducted from the dependency path
-        let { name, version } = versionData;
-        if (!(name && version)) {
-          ({ name, version } = this.parseDepPath(depPath));
+        const { name, version } = versionData;
+        let parsedPath: ParsedDepPath = {};
+        if (!(version && name)) {
+          parsedPath = this.parseDepPath(depPath);
         }
+
         const pkg: NormalisedPnpmPkg = {
           id: depPath,
-          name,
-          version,
+          name: name || parsedPath.name,
+          version: version || parsedPath.version || depPath,
           isDev: versionData.dev == 'true',
-          dependencies: versionData.dependencies,
-          devDependencies: versionData.devDependencies,
-          optionalDependencies: versionData.optionalDependencies,
+          dependencies: versionData.dependencies || {},
+          devDependencies: versionData.devDependencies || {},
+          optionalDependencies: versionData.optionalDependencies || {},
         };
         packages[`${pkg.name}@${pkg.version}`] = pkg;
       },
     );
-    return packages;
+    this.extractedPackages = packages;
```

Comment:
Why are we doing it like this? Functions with side-effects looks a bit bad to me. But maybe I don't understand it.

Reply:
becasue the new lockfile version specifies dependencies and other important data in the 'snapshots' field that's being populated in the subclass constructor, after calling the super constructor. But yes, it's better to return the value and assign it where I call the function.

Reply:
fixed

Reply:
<img src="https://media0.giphy.com/media/GVMhZwYv8U5NK/giphy.gif"/>